dotCMS/src/main/java/org/apache/felix/framework/OSGIUtil.java

```diff
@@ -257,6 +257,73 @@ public String getOsgiExtraConfigPath () {
         return Paths.get(dirPath).normalize().toString();
     }
 
+    /**
+     * Test if the Osgi Packages are ok to be overriden
+     * @param osgiPackages
+     */
+    public void testDryRun (final String osgiPackages) {
+
+        final List<ParsedHeaderClause> exportClauses =
+                invokeParserStandardHeader(osgiPackages);
+        for (final ParsedHeaderClause clause : exportClauses) {
+
+            for (final String packageName : clause.m_paths) {
+                if (packageName.equals(".")) {
+
+                    Logger.error(this, "Exporing '.' is invalid.");
+                    throw new OsgiException("Exporing '.' is invalid.");
+                }
+
+                if (packageName.length() == 0) {
+
+                    Logger.error(this, "Exported package names cannot be zero length.\nPackages: " + osgiPackages);
+                    throw new OsgiException(
+                            "Exported package names cannot be zero length.");
+                }
+            }
+
+            // Check for "version" and "specification-version" attributes
+            // and verify they are the same if both are specified.
+            final Object versionAttr = clause.m_attrs.get(Constants.VERSION_ATTRIBUTE);
+            final Object packageSpecVersion = clause.m_attrs.get(Constants.PACKAGE_SPECIFICATION_VERSION);
+            if ((versionAttr != null) && (packageSpecVersion != null)) {
+                // Verify they are equal.
+                if (!String.class.cast (versionAttr).trim().equals(String.class.cast(packageSpecVersion).trim())) {
+                    throw new OsgiException(
+                            "Both version and specification-version are specified, but they are not equal.");
+                }
+            }
+
+            try {
+                if (versionAttr != null) {
+                    // check version format
+                    Version.parseVersion(versionAttr.toString());
+                }
+
+                if (packageSpecVersion != null) {
+                    // check version format
+                    Version.parseVersion(packageSpecVersion.toString());
+                }
+            } catch (IllegalArgumentException e) {
+
+                Logger.error(this, e.getMessage() + ".\nPackages: " + osgiPackages);
+                throw new OsgiException(e.getMessage(), e);
+            }
+        }
+    }
+
+    private  List<ParsedHeaderClause>  invokeParserStandardHeader (final String osgiPackages) {
+
+        try {
+            final Method method = ManifestParser.class.getDeclaredMethod("parseStandardHeader", String.class);
+            method.setAccessible(true);
```

Comment:
The approved way of doing this in Java21 is to use MethodHandles.   it is a little more complicated but something like this I got from ChatGPT refactor
```
private List<ParsedHeaderClause> invokeParserStandardHeader(final String osgiPackages) {
    try {
        // Lookup object for accessing private methods
        MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(ManifestParser.class, MethodHandles.lookup());
        
        // Define the method type (return type and parameter types)
        MethodType methodType = MethodType.methodType(List.class, String.class);
        
        // Lookup the method handle for parseStandardHeader
        MethodHandle methodHandle = lookup.findStatic(ManifestParser.class, "parseStandardHeader", methodType);
        
        // Invoke the method with the required parameter
        return (List<ParsedHeaderClause>) methodHandle.invoke(osgiPackages);
    } catch (Throwable e) {
        throw new OsgiException("Cannot access the parseStandardHeader to run the dry run", e);
    }
}
```

Reply:
Yeah I have end up doing this after a huge analysis of the OSGI code to avoid some of the export packages issues that the clients may be facing.
May be we should encapsulate the dyna call into a ReflectionUtils class and them move it to java 21 (as soon as we are not running just as a runtime)